Add an error when saving a record that was previously destroyed

arielj opened the pull request:
### Motivation / Background

Currently, if we call `save` or `save!` in an object that was already destroyed, `save` returns `false` and `save!` raises an exception, but there's no information about why the save failed.

I ran into this issue with a legacy app that, because of a complicated set of callbacks, was destroying an object at some point and then calling `save!` somewhere else. This produced an exception, but there was no information about the reason.

After a lot of research I realized that the issue was the record was destroyed (so `.destroyed?` was `true`), but I think it will help a lot to have an error added to avoid having to find this out.

### Detail

This PR updates the `create_or_update` method to add a `:base` error message for this specific scenario.

I added a configuration for this to keep the old behavior if needed and during upgrades, since I imagine some apps may rely on the errors hash being empty.

### Additional information

No additional information.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

nvasilevski reviewed (approved):
This seems like a good addition! Not sure if the core team will consider the configuration an overkill but it also can't hurt to have it (except for eventual deprecation)

akhilgkrishnan reviewed (commented):
@arielj You need to add an entry for new settings in [Default Values for Target Version 7.1](https://github.com/rails/rails/blob/main/guides/source/configuring.md#default-values-for-target-version-71) to fix the lint issue.

```diff
diff --git a/guides/source/configuring.md b/guides/source/configuring.md
index 0bc3253ae1..b6d039581f 100644
--- a/guides/source/configuring.md
+++ b/guides/source/configuring.md
@@ -73,6 +73,7 @@ Below are the default values associated with each target version. In cases of co
 - [`config.active_record.default_column_serializer`](#config-active-record-default-column-serializer): `nil`
 - [`config.active_record.encryption.hash_digest_class`](#config-active-record-encryption-hash-digest-class): `OpenSSL::Digest::SHA256`
 - [`config.active_record.encryption.support_sha1_for_non_deterministic_encryption`](#config-active-record-encryption-support-sha1-for-non-deterministic-encryption): `false`
+- [`config.active_record.error_saving_destroyed`](#config-active-record-error-saving-destroyed): `true`
 - [`config.active_record.generate_secure_token_on`](#config-active-record-generate-secure-token-on): `:initialize`
 - [`config.active_record.marshalling_format_version`](#config-active-record-marshalling-format-version): `7.1`
 - [`config.active_record.query_log_tags_format`](#config-active-record-query-log-tags-format): `:sqlcommenter`
```

arielj committed:
Add an error when saving a record that was previously destroyed

Before this change, if we call `save` or `save!` in an object that was already destroyed,
`save` returns `false` and `save!` raises an exception, but there's no information about
why the save actually failed.

This adds a `:base` error message in this scenario.

This also adds a configuration to be able to turn this on eventually when upgrading

zzak commented:
Just a couple of thoughts, in no particular order:

* `false` is the correct API here
* I think we generally don't use validations/errors for internal API like this, exceptions are preferred
* `save!` variations already raise an exception, perhaps it could be more specific?
* The name `raise_when_saving_already_destroyed` is more understandable

Probably getting confirmation on that errors vs. exception context is a good place to ask in [discord](https://rubyonrails.org/2022/6/13/rails-discord-server-is-now-open-to-the-public).
